### PR TITLE
Index documents by user ID instead of username.

### PR DESCRIPTION
**Pull request by dbutenhof**
Whether users are identified primarily by username or email, we want all external forms of identification to be changeable; so we will index Elasticsearch documents using a stable internal identification, the user's table row ID.

This removes the user creation hack from `pbench-state-manager` and utilizes the `pbench-user-create` tool to create users referenced in each test's `.db` file along with a set of "standard" users required by some of the satellite and prep-shim tests.

A number of CLI unit test gold files change in fairly obvious ways.

### CODE REVIEW

**Review by webbnh — APPROVED** [review]
Below is the usual pile of nits and suggestions.

I'm marking this "Approved", but I really don't like the way the unit test script is invoking `user_create.py`:  I think it should be invoking it via the `pbench-user-create` command instead.  It seems like the Click-based commands should be available to the unit test scripts; and, if they are not, how can we be assured that all of the requirements for the rest of the Python code are present?  (And, if we _do_ have that assurance, then it should be straightforward to have the Click commands available as well.)

File: server/bin/gold/test-7.21.txt
Comment (dbutenhof) [reply server/bin/gold/test-7.21.txt:103]: `test-7.21.db` defines two datasets, one with username `fake` and the other with username `test`; they're sorted, so `fake` gets number 1 and `test` gets number 2. The datasets are defined in the `.db` file in reverse chronological order, so the "second" gets indexed first. Weird, but all perfectly reasonable. ;-)

File: server/bin/unittests
Comment (dbutenhof) [reply server/bin/unittests:267]: It's the pragmatic straightforward way to do this, though!

Yeah, it might be "fun" to make the `unittests` script do a local install of `pbench-server` under its `/var/tmp` path and use that to allow us to access the full click-ified commands; but all we need for the "traditional" commands is `PYTHONPATH` and `PATH`, which we have... so that's a substantial amount of new work to make this one bit look a bit cleaner. I decided that effort wasn't justifiable right now.

Comment (dbutenhof) [reply server/bin/unittests:267]: It was a cute little bash-ism for the user's "real name": it capitalizes the substitution!

```
dbutenho 16:44 templatedb:templatedb x="david"
dbutenho 17:03 templatedb:templatedb echo ${x^}
David
```

File: lib/pbench/server/database/models/tracker.py
Comment (webbnh) [reply lib/pbench/server/database/models/tracker.py:321]: I'm glad that you were prepared....  :-)  Did something change?  :-D

Datasets have to be owned by _someone_, right?  So, we cannot delete the owner out from under them without either deleting them as well or attaching them to a new owner.  (I'm not sure that we should consider public datasets differently from private datasets -- yes, we probably don't want to delete public datasets, but that should be addressed by having a new owner "adopt" them somehow.)

I kind of like the idea of refusing to delete a user who owns any datasets (), but that breeds the question of what to do when the (human) user is gone and there's no obvious avenue for getting his/her datasets deleted?

For now, I guess we just delete the user out from under them and orphan the datasets -- there's nothing that says that an orphaned dataset cannot be adopted (or deleted) later, right?  And, the orphans will still be associated with their original user's ID, it's just that there won't be a `User` associated with that ID, which, hopefully, won't cause any problems with normal operations.

File: lib/pbench/server/database/models/tracker.py
Comment (webbnh) [reply lib/pbench/server/database/models/tracker.py:363]: (I'm really looking forward to seeing what you come up with for hints...I'm hoping that you don't punt and use `Any` for the return value....  :-) )

File: lib/pbench/test/unit/server/conftest.py
Comment (webbnh) [reply lib/pbench/test/unit/server/conftest.py:131]: As a reviewer, I mostly appreciate that.  :-D  But, since this function refers to that one, it seems likely that that one should have the complementary explanation referring to this one.

File: lib/pbench/test/unit/server/test_state_tracker.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_state_tracker.py:66]: I realize that it makes for monotonous coding (and yucky reviewing), but I think it's best to have each test scenario test one, single, well-defined aspect.  If it works out right, then fewer tests fail when something goes wrong, and the result is easier to triage.

File: lib/pbench/test/unit/server/test_state_tracker.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_state_tracker.py:46]: And this way, if someone decides to make a change to the contents of `create_user`, then they don't have to make the corresponding change to each and every test case....

File: server/bin/gold/test-7.21.txt
Comment (webbnh) [reply server/bin/gold/test-7.21.txt:103]: ["Who is Number 1?"](https://www.quotes.net/mquote/918218)

File: lib/pbench/server/database/models/tracker.py
Comment (dbutenhof) [reply lib/pbench/server/database/models/tracker.py:363]: It's awkward, because `value` is flexible; after all, we're here to validate it as an unknown. But if we return anything, it's a `User`. I'm debating, though, whether it's `value: str` or `value: Any`...

The thing is, I feel a bit weird adding type hints just to this new code when they're missing from most of the rest of the APIs; but I don't want to add them all, either.

File: server/bin/unittests
Comment (webbnh) [reply server/bin/unittests:267]: > all we need for the "traditional" commands is `PYTHONPATH` and `PATH`

That and all the stuff in the `requirements.txt` and/or `setup.py`...which, by the way, is all that is needed to set up the Click commands.

(I think a `pip install .` will do it...I'm not sure what that "it" is -- I do it in a Python "virtual environment", but I assume that the unit tests are already doing something similar -- but, after issuing that command, then the Click tools are immediately accessible via my shell path.)

File: server/bin/gold/test-7.21.txt
Comment (dbutenhof) [reply server/bin/gold/test-7.21.txt:103]: I sometimes wonder how that would age if I watched an episode ... if that's even possible.

File: server/bin/unittests
Comment (webbnh) [reply server/bin/unittests:267]: I figured it was something like that.  B-)

I had to ask because it didn't work in the bash I have on OS X:
```
wscales-mac:pbench_upstream wscales$ foo=dave
wscales-mac:pbench_upstream wscales$ echo ${foo^}
-bash: ${foo^}: bad substitution
```
But, it works in my Centos-8 container.  :-D

File: lib/pbench/server/database/models/tracker.py
Comment (webbnh) [reply lib/pbench/server/database/models/tracker.py:363]: Oh, I'm pretty sure that `value` is an `Any`.  It's the _return_ value that I want to see you describe!  (I assume it's a union of `User` and `str`, but I'm waiting to see what you come up with.)

As for adding type hints, we should not skip them for new code, and we'll get to the existing code when it seems appropriate.

File: server/bin/gold/test-7.21.txt
Comment (webbnh) [reply server/bin/gold/test-7.21.txt:103]: Amazon Prime has, I think, all of the episodes.  I watched them when they were up for free ([They are up with ads, presently](https://www.amazon.com/Arrival/dp/B07F24RGLC/ref=sr_1_1?crid=2V2XN6N6ONCHJ)).  The show is clearly a period piece (the fashions are fabulous), but I thought it bore up well.  It's wonderfully weird, but I do enjoy a little surrealism now and again.

**Review by webbnh — APPROVED** [review]
GTG.

File: server/bin/unittests
Comment (webbnh) [inline server/bin/unittests:482]: Alternatively,
```
    local users = "pbench quarantiner satellite"
    if [ -f ${_state_db} ]; then
        users="$(cat ${_state_db} | cut --d ' ' -f 4 | sort -u) ${users}"
    fi
```

**Review by webbnh — DISMISSED** [review]
I find the enthusiasm in the commit log for your latest update to be grossly underwhelming.  Let me help with that:  Whoopee!!!!!  🥇 

I think enabling and using the Click command is much better.  :-)  Thank you.

However, since you made me look, I found something to suggest.  ;-)

**Comment by webbnh** [comment]
> This pull request **introduces 1 alert** when merging [654ca22](https://github.com/distributed-system-analysis/pbench/commit/654ca22a92057f7e76b7345d43d9f6a49f0a2d9a) into [5fdd398](https://github.com/distributed-system-analysis/pbench/commit/5fdd398048adc9a0a3365e6abcb476a82663a1ab) - [view on LGTM.com](https://lgtm.com/projects/g/distributed-system-analysis/pbench/rev/pr-8ba800644e9f591100df878d17f296329c00db54)
> 
> **new alerts:**
> 
> * 1 for Unused import

@dbutenhof, I presume that this warning needs to be resolved before this PR can be merged...can you tell me again what is required to make that happen?  (Are we going to be able to do that before this week's Jam Session with Nikhil out?)

File: lib/pbench/cli/server/user_create.py
Comment (webbnh) [reply lib/pbench/cli/server/user_create.py]: Dave was able to make this happen.  (Hurray!!!)

File: server/bin/unittests
Comment (dbutenhof) [reply server/bin/unittests:482]: Well, you're not wrong, although I'm going to tire of re-running the whole CLI gold suite repeatedly. ;-)

Comment (webbnh) [reply server/bin/unittests:482]: And I'll eventually tire of reviewing the same PR repeatedly....  ;-)

(It's only a suggestion -- you've got your "Ship it!" from me -- but it _is_ nicer....  :-D )

**Comment by dbutenhof** [comment]
> > This pull request **introduces 1 alert** when merging [654ca22](https://github.com/distributed-system-analysis/pbench/commit/654ca22a92057f7e76b7345d43d9f6a49f0a2d9a) into [5fdd398](https://github.com/distributed-system-analysis/pbench/commit/5fdd398048adc9a0a3365e6abcb476a82663a1ab) - [view on LGTM.com](https://lgtm.com/projects/g/distributed-system-analysis/pbench/rev/pr-8ba800644e9f591100df878d17f296329c00db54)
> > **new alerts:**
> > 
> > * 1 for Unused import
> 
> @dbutenhof, I presume that this warning needs to be resolved before this PR can be merged...can you tell me again what is required to make that happen? (Are we going to be able to do that before this week's Jam Session with Nikhil out?)

This is interesting, because it's complaining about an unreferenced import despite the `# noqa F401` tag, which does the job for a local `flake8` run and also seemed to satisfy LGTM in Nikhil's #2263 where he also added imports for `db_init` that weren't directly referenced. I have no idea why this is happening, and if we're gating merge on LGTM this might show up after I have two approvals. I could hack it by writing an unused method that references `ActiveTokens`, but I'd rather not.

Perhaps Nikhil's change will avoid the necessity of pulling it in in this context; but with my reference to `User` inside of `Dataset` things were blowing up resolving `User`'s quoted backlink `"ActiveTokens"` reference and if there's a better solution I wasn't able to find it.

**Review by webbnh — DISMISSED** [review]
Yay!

**Review by webbnh — APPROVED** [review]

**Review by webbnh — DISMISSED** [review]
GitHub says "no change", it must be good!

File: lib/pbench/server/database/database.py
Comment (portante) [inline lib/pbench/server/database/database.py:25]: Perhaps make the message a common variable?

**Review by portante — CHANGES_REQUESTED** [review]
Needs a rebase and conflict resolution.

**Review by portante — DISMISSED** [review]

**Review by portante — APPROVED** [review]

**Comment by dbutenhof** [comment]
I'm hoping I've figured out the LGTM/flake8 thing ... Looks like LGTM wants the specific diagnostic in square brackets immediately following either "lgtm" or "noqa". So while flake8 is happy with `# noqa F401` LGTM expects either `# lgtm[F401]` or `# noqa[F401]`. I've verified that flake8 is happy with "noqa[F401]" so I'm hoping that LGTM will be, as well.

**Comment by dbutenhof** [comment]
> This pull request **introduces 1 alert** when merging [dce36da](https://github.com/distributed-system-analysis/pbench/commit/dce36da5cd37f741ffea77caaec9ea1ab31c1cd3) into [8286a17](https://github.com/distributed-system-analysis/pbench/commit/8286a1735ccd2152999bf938424450b099ca25a9) - [view on LGTM.com](https://lgtm.com/projects/g/distributed-system-analysis/pbench/rev/pr-906f45d25a6f1a7769da86afa22a6962b555a809)
> 
> **new alerts:**
> 
> * 1 for Unused import

OK, well clearly LGTM isn't happy with `# noqa[F401]` either, although the documentation suggests it should be. :-(

File: lib/pbench/server/database/database.py
Comment (webbnh) [inline lib/pbench/server/database/database.py:23]: I don't want to gate the jam session, but, since we seem to be investing some effort into this, it seems like, instead of printing directly to `stderr` when `logger` is absent, we should be using a default logger (which will presumably print to `stderr`).

I haven't done the full investigation, but if we just invoke the "root" logger, then I think we'll get the ["logger of last resort"](https://docs.python.org/3/library/logging.html#logging.lastResort), which will do what we want...I think.

Comment (webbnh) [reply lib/pbench/server/database/database.py:23]: That is, if `logger` is `None`, I think we can just instantiate a default instance of the `logging` class and use that instead, and then in common code issue the `error()` method call.

**Review by webbnh — APPROVED** [review]

**Pull request merged**
